cudax/include/cuda/experimental/__cuco/__hyperloglog/hyperloglog_impl.cuh

```diff
@@ -0,0 +1,604 @@
+    auto constexpr __block_size = 1024;
+    ::cuda::experimental::cuco::__hyperloglog_ns::__clear<<<1, __block_size, 0, __stream.get()>>>(*this);
+  }
+
+  //! @brief Adds an item to the estimator.
+  //!
+  //! @param item The item to be counted
+  _CCCL_DEVICE constexpr void __add(const _Tp& __item) noexcept
+  {
+    const auto __h      = __hash(__item);
+    const auto __reg    = __h & __register_mask;
+    const auto __zeroes = ::cuda::std::countl_zero(__h | __register_mask) + 1; // __clz
+
+    // reversed order (same one as Spark uses)
+    // const auto __reg    = __h >> ((sizeof(__hash_value_type) * 8) - __precision);
+    // const auto __zeroes = ::cuda::std::countl_zero(__h << __precision) + 1;
+
+    this->__update_max(__reg, __zeroes);
+  }
+
+  //! @brief Asynchronously adds to be counted items to the estimator.
+  //!
+  //! @tparam _InputIt Device accessible random access input iterator where
+  //! <tt>std::is_convertible<std::iterator_traits<_InputIt>::value_type,
+  //! _Tp></tt> is `true`
+  //!
+  //! @param __first Beginning of the sequence of items
+  //! @param __last End of the sequence of items
+  //! @param __stream CUDA stream this operation is executed in
+  template <class _InputIt>
+  _CCCL_HOST constexpr void __add_async(_InputIt __first, _InputIt __last, ::cuda::stream_ref __stream)
+  {
+    const auto __num_items = ::cuda::std::distance(__first, __last);
+    if (__num_items == 0)
+    {
+      return;
+    }
+
+    int __grid_size         = 0;
+    int __block_size        = 0;
+    const int __shmem_bytes = __sketch_bytes();
+    const void* __kernel    = nullptr;
+
+    // In case the input iterator represents a contiguous memory segment we can employ efficient
+    // vectorized loads
+    if constexpr (::cuda::std::contiguous_iterator<_InputIt>)
+    {
+      const auto __ptr                  = thrust::raw_pointer_cast(&__first[0]);
+      auto constexpr __max_vector_bytes = 32;
+      const auto __alignment =
+        1u << ::cuda::std::countr_zero(reinterpret_cast<::cuda::std::uintptr_t>(__ptr) | __max_vector_bytes);
+      const auto __vector_size = __alignment / sizeof(value_type);
+
+      switch (__vector_size)
+      {
+        case 2:
+          __kernel = reinterpret_cast<const void*>(
+            ::cuda::experimental::cuco::__hyperloglog_ns::__add_shmem_vectorized<2, _HyperLogLog_Impl>);
+          break;
+        case 4:
+          __kernel = reinterpret_cast<const void*>(
+            ::cuda::experimental::cuco::__hyperloglog_ns::__add_shmem_vectorized<4, _HyperLogLog_Impl>);
+          break;
+        case 8:
+          __kernel = reinterpret_cast<const void*>(
+            ::cuda::experimental::cuco::__hyperloglog_ns::__add_shmem_vectorized<8, _HyperLogLog_Impl>);
+          break;
+        case 16:
+          __kernel = reinterpret_cast<const void*>(
+            ::cuda::experimental::cuco::__hyperloglog_ns::__add_shmem_vectorized<16, _HyperLogLog_Impl>);
+          break;
+      };
+    }
```

Comment:
This whole dispatching yada yada could be replaced by a simple `cub::DeviceTransform` which would additionally utilize TMA copies when possible. However, since the kernel uses shared memory for the HLL sketch, this is not possible.